controllers/apply-next/simple/form.test.js

```diff
@@ -0,0 +1,892 @@
+
+        const mockWithCompanyNumber = mockFullForm({
+            country: 'england',
+            organisationType: 'not-for-profit-company',
+            companyNumber: '123456789'
+        });
+        expect(validate(mockWithCompanyNumber).error).toBeNull();
+    });
+
+    test('validate full form with charity number', () => {
+        const mockWithoutCharityNumber = mockFullForm({
+            country: 'england',
+            organisationType: 'not-for-profit-company'
+        });
+
+        expect(validate(mockWithoutCharityNumber).error).toBeInstanceOf(Error);
+
+        const mockWithCharityNumber = mockFullForm({
+            country: 'england',
+            organisationType: 'unincorporated-registered-charity',
+            charityNumber: '123456789'
+        });
+
+        expect(validate(mockWithCharityNumber).error).toBeNull();
+    });
+
+    test('return conditional role choices based on organisation type', () => {
+        const mappings = {
+            '': [],
+            'unregistered-vco': ['chair', 'vice-chair', 'secretary', 'treasurer'],
+            'unincorporated-registered-charity': ['trustee'],
+            'charitable-incorporated-organisation': ['trustee'],
+            'not-for-profit-company': ['company-director', 'company-secretary'],
+            'school': ['head-teacher', 'chancellor', 'vice-chancellor'],
+            'statutory-body': ['parish-clerk', 'chief-executive']
+        };
+
+        map(mappings, (expected, type) => {
+            const form = formBuilder({
+                locale: 'en',
+                data: { 'organisation-type': type }
+            });
+            expect(form.allFields.seniorContactRole.options.map(o => o.value)).toEqual(expected);
+        });
+    });
+
+    function fieldNamesFor(sectionSlug, stepTitle) {
+        return function(data) {
+            const form = formBuilder({
+                locale: 'en',
+                data: data
+            });
+
+            const section = form.sections.find(s => s.slug === sectionSlug);
+            const step = section.steps.find(s => s.title === stepTitle);
+            return map(flatMap(step.fieldsets, 'fields'), f => f.name);
+        };
+    }
+
+    test('registration numbers shown based on organisation type', () => {
+        const fieldNamesFn = fieldNamesFor('organisation', 'Registration numbers');
+
+        expect(fieldNamesFn({})).toEqual([]);
+
+        const mappings = {
+            'unincorporated-registered-charity': ['charity-number'],
+            'charitable-incorporated-organisation': ['charity-number'],
+            'not-for-profit-company': ['company-number', 'charity-number']
+        };
+
+        map(mappings, (expected, type) => {
+            const result = fieldNamesFn({ 'organisation-type': type });
+            expect(result).toEqual(expected);
+        });
+    });
+
+    test('exclude address and date of birth for schools or statutory bodies', () => {
```

Comment:
These tests were previously handled in multiple places via the `shouldShow` method. Now we are able to dynamically return fields based on form data directly in the model we can test this logic for the whole step, rather than only for the discreet fields. Gives us better guarantees that we're validating what is actually show, and is hopefully easier to understand what the rules are.